rmf_demo_plugins/rmf_plugins_common/include/rmf_plugins_common/readonly_common.hpp

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright (C) 2020 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+#ifndef RMF_PLUGINS_COMMON__READONLY_COMMON_HPP
+#define RMF_PLUGINS_COMMON__READONLY_COMMON_HPP
+
+#include <rclcpp/rclcpp.hpp>
+#include <rclcpp/logger.hpp>
+
+#include <memory>
+#include <unordered_map>
+#include <unordered_set>
+#include <mutex>
+#include <string>
+
+#include <Eigen/Geometry>
+
+#include <rmf_fleet_msgs/msg/robot_state.hpp>
+#include <building_map_msgs/msg/building_map.hpp>
+#include <building_map_msgs/msg/level.hpp>
+#include <building_map_msgs/msg/graph.hpp>
+
+namespace rmf_readonly_common {
+
+class ReadonlyCommon
+{
+public:
+
+  using BuildingMap = building_map_msgs::msg::BuildingMap;
+  using Level = building_map_msgs::msg::Level;
+  using Graph = building_map_msgs::msg::Graph;
+  using Location = rmf_fleet_msgs::msg::Location;
+  using Path = std::vector<Location>;
+
+  rclcpp::Node::SharedPtr ros_node;
+  std::string name;
+  Eigen::Isometry3d pose;
+  double sim_time = 0.0;
+
+  rclcpp::Logger logger();
+  template<typename SdfPtrT>
+  void read_sdf(SdfPtrT& sdf);
+  void init(rclcpp::Node::SharedPtr node);
+  void on_update();
+
+private:
+  rclcpp::Publisher<rmf_fleet_msgs::msg::RobotState>::SharedPtr _robot_state_pub;
+  rclcpp::Subscription<BuildingMap>::SharedPtr _building_map_sub;
+
+  rmf_fleet_msgs::msg::RobotState _robot_state_msg;
+  rmf_fleet_msgs::msg::RobotMode _current_mode;
+
+  bool _found_level = false;
+  bool _found_graph = false;
+  bool _initialized_graph = false;
+  bool _initialized_start = false;
+
+  // Store cache of BuildingMap
+  // BuildingMap _map;
+  Path _path;
+  Level _level;
+  Graph _graph;
+
+  std::string _level_name = "L1";
+  std::size_t _nav_graph_index = 1;
+  std::string _start_wp_name = "caddy";
+  // The number of waypoints to add to the predicted path. Minimum is 1.
+  std::size_t _lookahead = 1;
+  std::size_t _start_wp;
+  std::vector<std::size_t> _next_wp;
+
+  std::unordered_map<std::size_t,
+    std::unordered_set<std::size_t>> _neighbor_map;
+
+  double _last_update_time = 0.0;
+  double _update_threshold = 0.5; // Update every 0.5s
+  double _waypoint_threshold = 2.0;
+
+  bool _merge_lane = false;
+  double _lane_threshold = 0.2; // meters
+
+  int _update_count = 0;
+  std::string _current_task_id;
+
+  std::mutex _mutex;
```

Comment:
Nit, I'd try to name the mutex to something more explicative relevant to what it is used for